core/azure-core/src/main/java/com/azure/core/polling/Poller.java

```diff
@@ -0,0 +1,282 @@
+     * @param cancelOperation cancel operation
+     * @throws NullPointerException If {@code pollerOptions} is {@code null}.
+     * @throws NullPointerException If {@code pollOperation} is {@code null}.
+     */
+    public Poller(PollerOptions pollerOptions, Function<PollResponse<T>, Mono<PollResponse<T>>> pollOperation, Consumer<Poller> cancelOperation) {
+        this(pollerOptions, pollOperation);
+        this.cancelOperation = cancelOperation;
+    }
+
+    /**
+     * Calls cancelOperation function if provided.
+     * It will only call cancelOperation if {@link com.azure.core.polling.PollResponse.OperationStatus} is IN_PROGRESS otherwise does nothing.
+     *
+     * @throws UnsupportedOperationException when cancel operation is not provided.
+     */
+    public void cancelOperation() throws UnsupportedOperationException {
+        if (cancelOperation == null) {
+            throw new UnsupportedOperationException("Cancel operation is not supported on this service/resource.");
+        }
+
+        // We can not cancel an operation if it was never started
+        // It only make sense to call cancel operation if current status IN_PROGRESS.
+        if (pollResponse != null && pollResponse.getStatus() != PollResponse.OperationStatus.IN_PROGRESS) {
+            return;
+        }
+        //Time to call cancel
+        cancelOperation.accept(this);
+    }
+
+    /**
+     * Enable user to subscribe and listen on all the poll responses.
+     * The user will start receiving PollResponse when client subscribe to this Flux.
+     * The poller could still have its own auto polling in action unless user has turned off
+     * auto polling.
+     *
+     * @return poll response as Flux that can be subscribed.
+     */
+    public Flux<PollResponse<T>> getObserver() {
+        return fluxHandle;
+    }
+
+    /**
+     * Enable user to take control of polling and trigger manual poll operation. It will perform one call to poll operation.
+     * This will not turn off auto polling.
+     *
+     * @return a Mono of {@link PollResponse}
+     */
+    public Mono<PollResponse<T>> poll() {
+        return pollOperation.apply(pollResponse)
+            .doOnEach(pollResponseSignal -> {
+                if (pollResponseSignal.get() != null) {
+                    pollResponse = pollResponseSignal.get();
+                }
+            });
+    }
+
+    /**
+     * Blocks execution and wait for polling to complete.
+     * Auto polling must be turned on for poller to continuously poll in background.
+     *
+     * @return returns last poll response when polling is done.
+     */
+    public PollResponse<T> block() {
+        return fluxHandle.blockLast();
+    }
+
+    /*
+     * This function will apply delay and call poll operation function async.
+     * @return mono of poll response
+     */
+    private Mono<PollResponse<T>> asyncPollRequestWithDelay() {
+        return  Mono.defer(() -> pollOperation.apply(pollResponse)
```

Comment:
For my own education, why is `Mono.defer(..)` necessary here?

Reply:
If I do not use defer here the repeat() , see below,  does not work and I only see one time invocation 
        fluxHandle = asyncPollRequestWithDelay()
            .flux()
            .repeat()
            .takeUntil(pollResponse -> pollResponse != null && pollResponse.isDone())
            .share();